### PR TITLE
New rule: prefer_async_await_over_raw_futures

### PR DESCRIPTION
**Pull request by JPaulsen**
Fixes https://github.com/dart-lang/linter/issues/525

### CODE REVIEW

**Comment by JPaulsen** [comment]
I had to add this line:

```dart
Future<T> then(Future<T> onValue(T value), { Function onError });
```

But the original was:

```dart
Future<S> then<S>(FutureOr<S> onValue(T value), { Function onError });
```

File: lib/src/rules/prefer_async_await_over_raw_futures.dart [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+bool _isDynamic(DartType dartType) => dartType.name == 'dynamic';
+
+bool _isFunctionExpression(AstNode node) => node is FunctionExpression;
+
+bool _isRelevantMethodInvocation(ReturnStatement returnStatement) {
+  final expression = returnStatement.expression;
+  return (expression is MethodInvocation &&
+      (expression.methodName?.name == 'then' ||
+          expression.methodName?.name == 'catchError' ||
+          expression.methodName?.name == 'whenComplete') &&
```
Comment (bwilkerson) [inline lib/src/rules/prefer_async_await_over_raw_futures.dart]: nit: extract the name once and use a local variable

File: test/mock_sdk.dart [hunk 2]
```diff
@@ -160,6 +160,10 @@ class Future<T> {
   factory Future.delayed(Duration duration, [T computation()]) => null;
   factory Future.value([value]) => null;
   static Future wait(List<Future> futures) => null;
+  Future<T> catchError(Function onError, {bool test(Object error)});
+  Future<T> whenComplete(action());
+  Future<T> then(Future<T> onValue(T value), { Function onError });
+  // Original is Future<S> then<S>(FutureOr<S> onValue(T value), { Function onError });
```
Comment (bwilkerson) [inline test/mock_sdk.dart]: Why did you need to change this? We need to put the original back because it matches what's in the SDK, which is what every user will be analyzed against.

Comment (JPaulsen) [inline test/mock_sdk.dart]: Because there was nothing, the 3 methods did not appear, so the tests did not pass, so I had to add those 3 methods just like are in the SDK except the 3rd one.

Is there other way to do this?

File: lib/src/rules/prefer_async_await_over_raw_futures.dart (hunk 1 above)
Comment (JPaulsen) [inline lib/src/rules/prefer_async_await_over_raw_futures.dart]: Fixed

File: test/mock_sdk.dart (hunk 2 above)
Comment (bwilkerson) [inline test/mock_sdk.dart]: I was confused and thought that you had changed from the "original" definition to the new (even though you clearly added the commented out line a well). Adding those three methods is fine, but the definitions should match what is in the SDK. If that means that we also need to add the definition of `FutureOr`, then we should do so.

Comment (pq) [inline test/mock_sdk.dart]: +1 on making the mock as close to what we have in the SDK.  Feel free to add whatever you need (`FutureOr`, etc).  Thanks!

Comment (JPaulsen) [inline test/mock_sdk.dart]: Sorry for my confusing answer, my bad, I got it, I will try to fix this now, thanks to you both :)

**Comment by JPaulsen** [comment]
@pq @bwilkerson I changed it to:

```dart
Future<T> then(FutureOr<T> onValue(T value), { Function onError });
```

The problem appears when I add a new generic (S) to the then method:

```dart
Future<S> then<S>(FutureOr<S> onValue(T value), { Function onError });
```

Do you know how can I solve this?

The class FutureOr<T> was already defined in the file.

```dart
class FutureOr<T> {}
```

**Comment by bwilkerson** [comment]
Not off hand; if we can analyze the method in the SDK we should be able to analyze it anywhere.

Can you give me more details about what problem you're seeing (an exception, an unexpected error code, etc.)?

**Comment by JPaulsen** [comment]
pub run test -N prefer_async_await_over_raw_futures

```
00:01 +0: test/rule_test.dart: rule dart prefer_async_await_over_raw_futures                       

/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 10:35 [static type warning] 'then' is not a method
  return longRunningCalculation().then((result) {
                                  ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 18:35 [static type warning] 'then' is not a method
  return longRunningCalculation().then((result) {
                                  ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 30:35 [static type warning] The method 'timeout' is not defined for the class 'Future'.
  return longRunningCalculation().timeout(null, onTimeout: () {
                                  ^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 36:35 [static type warning] 'then' is not a method
  return longRunningCalculation().then((result) {
                                  ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 44:35 [static type warning] 'then' is not a method
  return longRunningCalculation().then((result) {
                                  ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 56:35 [static type warning] The method 'timeout' is not defined for the class 'Future'.
  return longRunningCalculation().timeout(null, onTimeout: () {
                                  ^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 63:37 [static type warning] 'then' is not a method
    return longRunningCalculation().then((result) {
                                    ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 71:37 [static type warning] 'then' is not a method
    return longRunningCalculation().then((result) {
                                    ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 83:37 [static type warning] The method 'timeout' is not defined for the class 'Future'.
    return longRunningCalculation().timeout(null, onTimeout: () {
                                    ^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 89:37 [static type warning] 'then' is not a method
    return longRunningCalculation().then((result) {
                                    ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 97:37 [static type warning] 'then' is not a method
    return longRunningCalculation().then((result) {
                                    ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 109:37 [static type warning] The method 'timeout' is not defined for the class 'Future'.
    return longRunningCalculation().timeout(null, onTimeout: () {
                                    ^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 23:12 [lint] Prefer async/await over using raw futures.
Future bad3() { // LINT
           ^^^^^^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 49:24 [lint] Prefer async/await over using raw futures.
dynamic badWithDynamic3() { // LINT
                       ^^^^^^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 76:3 [lint] Prefer async/await over using raw futures.
  Future bad3() { // LINT
  ^^^^^^^^^^^^^^^^^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 102:3 [lint] Prefer async/await over using raw futures.
  dynamic badWithDynamic3() { // LINT
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

null files analyzed, 16 issues found, in null ms.

00:01 +0 -1: test/rule_test.dart: rule dart prefer_async_await_over_raw_futures [E]                
  Expected: matches [Annotation:<[LINT]: "null" (line: 9) - [null:null]>, Annotation:<[LINT]: "null" (line: 17) - [null:null]>, Annotation:<[LINT]: "null" (line: 23) - [null:null]>, Annotation:<[LINT]: "null" (line: 35) - [null:null]>, Annotation:<[LINT]: "null" (line: 43) - [null:null]>, Annotation:<[LINT]: "null" (line: 49) - [null:null]>, Annotation:<[LINT]: "null" (line: 62) - [null:null]>, Annotation:<[LINT]: "null" (line: 70) - [null:null]>, Annotation:<[LINT]: "null" (line: 76) - [null:null]>, Annotation:<[LINT]: "null" (line: 88) - [null:null]>, Annotation:<[LINT]: "null" (line: 96) - [null:null]>, Annotation:<[LINT]: "null" (line: 102) - [null:null]>] unordered
    Actual: [
              Annotation:[LINT]: "Prefer async/await over using raw futures." (line: 49) - [24:103],
              Annotation:[LINT]: "Prefer async/await over using raw futures." (line: 76) - [3:122],
              Annotation:[LINT]: "Prefer async/await over using raw futures." (line: 102) - [3:134],
              Annotation:[LINT]: "Prefer async/await over using raw futures." (line: 23) - [12:103]
            ]
     Which: has too few elements (4 < 12)
  
  test/rule_test.dart 445:7  testRule.<fn>
  
00:01 +0 -1: Some tests failed.  
```

I got a lot of 'then is not a method' that is the critical part, in mock_sdk.dart I had:

```
Future<S> then<S>(FutureOr<S> onValue(T value), { Function onError });
```

**Comment by bwilkerson** [comment]
@pq Perhaps you'll know what's going on. I don't have any experience with "pub run test".

**Comment by bwilkerson** [comment]
Something just occurred to me. When you resolve dependencies, which version of analyzer are you using? It's possible that we need to update the dependencies such that we are guaranteed to get a version of analyzer that is new enough that it supports the generic method syntax.

**Comment by pq** [comment]
Along the lines of @bwilkerson's suggestion, maybe we can sidestep `pub` for the moment and see if things improve.

What happens if you try:

    dart test/util/rule_debug.dart prefer_async_await_over_raw_futures

?

**Comment by JPaulsen** [comment]
@pq 

```
$dart test/util/rule_debug.dart prefer_async_await_over_raw_futures
00:00 +0: prefer_async_await_over_raw_futures
00:00 +1: All tests passed!
```

**Comment by JPaulsen** [comment]
@pq same error code than above plus this:

```
00:00 +0 -1: Some tests failed.

Unhandled exception:
Dummy exception to set exit code.
#0      _rootHandleUncaughtError.<anonymous closure> (dart:async/zone.dart:1146)
#1      _microtaskLoop (dart:async/schedule_microtask.dart:41)
#2      _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
#3      _Timer._runTimers (dart:isolate-patch/timer_impl.dart:396)
#4      _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416)
#5      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
```

**Comment by JPaulsen** [comment]
This is using the line:

```dart
Future<S> then<S>(FutureOr<S> onValue(T value), { Function onError });
```
In mock_sdk.dart

**Comment by JPaulsen** [comment]
@bwilkerson I think so, we had this similar problem before https://github.com/dart-lang/linter/pull/484#pullrequestreview-26309119

**Comment by pq** [comment]
Where are we on this one?

Something that'd be nice to cleanup is error reporting.  Notice the inconsistency in what nodes are annotated with the errors below?

```
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 97:37 [static type warning] 'then' is not a method
    return longRunningCalculation().then((result) {
                                    ^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 109:37 [static type warning] The method 'timeout' is not defined for the class 'Future'.
    return longRunningCalculation().timeout(null, onTimeout: () {
                                    ^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 23:12 [lint] Prefer async/await over using raw futures.
Future bad3() { // LINT
           ^^^^^^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 49:24 [lint] Prefer async/await over using raw futures.
dynamic badWithDynamic3() { // LINT
                       ^^^^^^^^^^^^
/usr/local/google/home/casanueva/Development/linter/test/rules/prefer_async_await_over_raw_futures.dart 76:3 [lint] Prefer async/await over using raw futures.
  Future bad3() { // LINT

```

I think we want to get more specific with the node in the call `rule.reportLint(nodeToLint)`.

**Comment by JPaulsen** [comment]
@pq the node linted is all the function expression, because that must change to sync, and inside the body use await, the terminal only show the first line of the entire lint, but this lint in particular will use a lot of them, do you suggest a better node to lint?

**Comment by bwilkerson** [comment]
Could we flag the first invocation of one of the methods that causes the lint to be generated? In all four of the cases you displayed, I think the lint could be associated with the invocation of `whenComplete`.

**Comment by matanlurey** [comment]
@bwilkerson Author is no longer at Google. I think it's safe to close this PR.
